specs/emip-005.md

```diff
@@ -0,0 +1,472 @@
+
+2) Compute an ephemeral key pair using `ED255519 extended`
+1) Compute the shared DH secret between the private key from step (1) and the public key received (using the `exchange` functionality)
+1) Compute the `Blake2b512` hash of the secret
+1) Use this as the password to encrypt the password in (1) using [the Yoroi encryption spec](https://github.com/Emurgo/yoroi-frontend/blob/737595fec5a89409aacef827d356c9a1605515c0/docs/specs/code/ENCRYPT.md)
+
+The structure will look like the following:
+
+```
+COSE_Encrypt = [
+  Headers,
+  ciphertext : bstr / nil, # contained signed_message encrypted with random password
+  recipients : [+COSE_recipient]
+]
+
+COSE_recipient = [
+  Headers,
+  ciphertext : bstr / nil, # contains random password encrypted with shared secret
+]
+
+PubKeyEncryption = 96 (COSE_Encrypt)
+```
+
+The `Headers` for the body MUST have `enc_type: 2` in the `unprotected` field.
+
+The `Headers` for the recipient MUST have a `eph_pub: bstr` label containing the public key of the ephemeral keypair.
+
+## Multisig
+
+TODO
+
+## More headers
+
+### Version
+
+The `Headers` for the body MUST have `version: uint` in the `unprotected` field. See Changelog section for possible versions.
+
+### CWT
+
+CWT tags MAY be introduced inside `protected` headers as a CBOR map with the key `cwt`. It is RECOMMENDED to use these to avoid replay attacks
+
+### Payload encoding
+
+To solve `E3`, `signed_message` body header MUST contain `hashed: bool` as an `unprotected` header which defined whether or not we signed the `payload` OR the `Blake2b224` hash of the `payload`. The hash MUST be used in the following two cases
+
+1) The size of the raw `payload` would otherwise be too big to fit in hardware wallet memory (see E1). Note that the exact size for which this is the case depends on the device.
+1) The payload characters (ex: non-ASCII) that cannot be displayed on the hardware wallet device (see E3)
+
+We RECOMMEND showing the user the full payload on the device if possible because it lowers the attack surface (otherwise the user has to trust that the hash of the payload was calculated correctly).
+
+`Blake2b224` was chosen specifically because `224` bits is already a long string for hardware wallets.
+
+## User-Facing Encoding
+
+Once we have our top-level `encrypted_message` or `signed_message` we need to encode them in way that can be displayed to users (doesn't need to be stored and can be inferred just from the data)
+
+We define the encoding in three parts `prefix || data || checksum` (where || means append)
+
+### Prefix
+
+We need a human-readable prefix. We use "CM" for "Cardano Message" followed by the message type:
+
+- `encrypted_message`: `cme_`
+- `signed_message`: `cms_`
+- `mac_message`: `cmm_` (unused in this spec)
+
+### Data
+
+Data is simply the `base64url` encoding of the message
+
+### Checksum
+
+We use `CRC32` on the data for the checksum and store it as the `base64url` encoding of its network byte order representation.
```

Comment:
I used `CRC32` since we use it elsewhere but it seems CRC in general has fallen out of popularity. I'm open for a different checksum method if we want.

Reply:
After looking into this some more, I think `CRC32` is probably not a good choice even though it's already implemented inside HW wallets.

CRC32 has
1) Fairly few collisions compared to other hashes
2) Moderately slower than other hashes
3) Requires more memory than alternatives (need to build a lookup table)
4) Somewhat complex implementation (many alternatives exist)

Something like `fnv32a` would probably be better because
1) It still has fairly few collisions
2) Faster than CRC32 in general
3) Only O(1) memory requirement
4) Dead simple to implement

For general computers, it's not that important but (2) and (3) matter a lot for HW wallets since (3) frees up an additional 1024 bytes of memory